### ISSUE TITLE
Extend ignore private to also check the import path

### ISSUE THREAD

**Issue by KyleKing** [mkdocstrings/griffe#183]
**Describe the solution you'd like**

I wondered if the below logic could be extended to check not just the function name but also the import path for a `_` prefix:

https://github.com/mkdocstrings/griffe/blob/1e7b869b7345b30185e6bf634bc4f848c940172f/src/griffe/diff.py#L474-L476

To something like:

```python
if ignore_private and any(pth.startswith("_") for pth in [*old_obj.path.split('.'), name]):
    logger.debug(f"API check: {old_obj.path}.{name}: skip private object")
    continue
```

However, the issue with this naive approach is that `old_obj.path` is the definition path, when what I actually want is to check every possible path the object could be imported from to see if at least one is public. From my below example, I would want to treat `from module_name import public_thing` as public because the `module_name/__init__.py` includes `from ._private.things import public_thing`, but from my understanding, the `old_obj.path=_private.things.public_thing`, which would be treated as private by this initial implementation

**Context/Background.**

> TLDR: a human would consider this code an attempt to import from the private API: `from ._private.things import public_thing` because of `_private`, which I think should be how `griffe` works as well

I find that this is a pattern I often use to avoid redundancy where I can make a single underscore prefix to the file or folder name (example: https://github.com/KyleKing/pytest_cache_assert/tree/8ccc01d63a1df456896c1f88e25f2d8afc5831e0/pytest_cache_assert where `_check_assert/*` is private, but all imports from this `__init__.py` file should be considered public because they are re-exported: https://github.com/KyleKing/pytest_cache_assert/blob/8ccc01d63a1df456896c1f88e25f2d8afc5831e0/pytest_cache_assert/__init__.py)

The reason I like this approach is that I can later choose to make a module or file part of the public API just by renaming and updating the import statements rather than touching the function and class names everywhere they are used. 

There are other examples of this pattern, such as in `beartype`: https://github.com/beartype/beartype/tree/7d9dc4efd2b14867d80f292cc328f204c1d4c888/beartype, so I don't think I would be the only developer to benefit from this feature

**Describe alternatives you've considered**

I think this feature might be common enough to warrant a potentially breaking change to the core API behavior.

Alternatively or in coordination with the change to the default behavior, maybe the ignore private logic could be exposed to the Plugin API to allow for keeping the original behavior, extending the logic, or providing a way for user to configure their own rules for defining private code (such as manual ignores as discussed here: #131)

However, I would advocate for KISS and make the default behavior treat any possible import that can be included without referencing a string that is prefixed with an underscore as public

**Other.**

> Side note: this is just from a cursory glance, but is `{old_obj.path}.{name}` incorrect (i.e. would evaluate to `module.the_name.the_name`, but should be `module.the_name`)? It looks like the object path is the canonical path including the name, so maybe it should just be `{old_obj.path}` and my version would be: `if ignore_private and any(pth.startswith("_") for pth in old_obj.path.split('.')):`, right?
>
> https://github.com/mkdocstrings/griffe/blob/1e7b869b7345b30185e6bf634bc4f848c940172f/src/griffe/dataclasses.py#L648-L670

**Comment by KyleKing** [comment]
I might be wrong about my assumptions above and the ignore logic might be more complex elsewhere. Here is a repeatable example:

```sh
gh repo clone beartype/beartype
# install griffe, which could be with poetry like: `rm pyproject.toml && poetry init --no-interaction && poetry add griffe
poetry run python tmp.py
```

For `tmp.py` of:

```python
import griffe

previous = griffe.load_git("beartype", ref="v0.13.0")
current = griffe.load("beartype", ref="v0.15.0")

for breakage in griffe.find_breaking_changes(previous, current):
    print(breakage._explain_oneline())
```

Which outputs:

```
beartype/meta.py:92: PYTHON_VERSION_MIN: Attribute value was changed: '3.7.0' -> '3.8.0'
beartype/meta.py:211: VERSION: Attribute value was changed: '0.13.0' -> '0.15.1'
beartype/meta.py:294: URL_HOMEPAGE: Attribute value was changed: 'https://beartype.readthedocs.io' -> f'https://{PACKAGE_NAME}.readthedocs.io'
beartype/meta.py:322: URL_REPO: Attribute value was changed: 'https://github.com/beartype/beartype' -> f'https://github.com/{URL_REPO_ORG_NAME}/{URL_REPO_BASENAME}'
beartype/meta.py:382: LIBS_RUNTIME_OPTIONAL: Attribute value was changed: (f'typing_extensions >={_LIB_RUNTIME_OPTIONAL_VERSION_MINIMUM_TYPING_EXTENSIONS}') -> (f'typing-extensions >={_LIB_RUNTIME_OPTIONAL_VERSION_MINIMUM_TYPING_EXTENSIONS}')
beartype/meta.py:415: LIBS_TESTTIME_OPTIONAL: Attribute value was changed: ('mypy >=0.800; platform_python_implementation != "PyPy"', 'numpy; sys_platform != "darwin" and platform_python_implementation != "PyPy"', 'pandera', 'sphinx; python_version >= "3.8.0"', 'typing_extensions; python_version < "3.9.0"') -> ('mypy >=0.800; platform_python_implementation != "PyPy"', 'numpy; sys_platform != "darwin" and platform_python_implementation != "PyPy"', 'pandera', 'sphinx; python_version >= "3.8.0"', f'typing-extensions >={_LIB_RUNTIME_OPTIONAL_VERSION_MINIMUM_TYPING_EXTENSIONS}')
beartype/claw/_clawmain.py:209: beartype_package(package_names): Parameter was removed
beartype/claw/_clawmain.py:209: beartype_package(package_name): Parameter was added as required
beartype/roar/__init__.py:42: BeartypeClawRegistrationException: Public object was removed 
```

I'm surprised to not see more breaking changes (which might be because `beartype` uses more leading underscores that I do), but for `beartype/claw/_clawmain.py` both breaking changes appear correct. `beartype_package` is re-exported in `claw/__init__.py` :

```sh
> rg "beartype_package"
...

claw/__init__.py
20:#beartype_package() or beartype_packages() functions. After implementing that
32:    beartype_package as beartype_package,
33:    beartype_packages as beartype_packages,

...
```

**Comment by KyleKing** [comment]
For context, I'm experimenting with using `griffe` for tracking SemVer with the assumption that major versions are tracked manually

So far, I've got something like: https://github.com/KyleKing/calcipy/blob/2ed196707271453e8b8abc6678112edfccf6ed08/calcipy/experiments/bump_programmatically.py

**Comment by pawamoy** [comment]
(quick answer on mobile)

The current logic says that if a module is private we won't recurse into it. So if a public object from a private module is reported by Griffe, it's that this object was re-exported as public somewhere else. Is that the behavior you expect?

**Comment by KyleKing** [comment]
Excellent! That was what I was missing
